File: Contest #1/C++/G-Katryoshka.cpp
#include <iostream>
using namespace std;

/*
=> Body combination:
    1. Two eyes and one body.
    2. Two eyes, one mouth, and one body.
    3. One eye, one mouth, and one body. 
*/

int main()
{
    long long eyes, mouths, bodies, comp = 0;
    
    cin >> eyes >> mouths >> bodies;

    /*--------------------------------------------------------*/
    /* Wild cases */
    if (bodies == 0 || eyes == 0)
    {
        comp = 0;
        cout << comp << endl;
        return (0);
    }

    if (eyes == 1)
    {
        comp = 1;
        cout << comp << endl;
        return (0);
    }

    if (bodies == 1)
    {
        comp = 1;
        cout << comp << endl;
        return (0);
    }

    /* According to number of eyes */
    if (bodies >= eyes && mouths >= eyes)
    {
        comp += eyes;
        cout << comp << endl;
        return (0);
    }

    if (eyes >= bodies && mouths >= bodies)
    {
        comp += bodies;
        cout << comp << endl;
        return (0);
    }
    /*--------------------------------------------------------*/

    /* 1 eye, 1 body, 1 month */
    if (mouths > 0 && eyes >= mouths && bodies >= mouths)
    {
        comp += mouths;
        eyes -= comp;
        bodies -= comp;
        mouths = 0;
    }

    /* 2 eye, 1 body */
    if (bodies >= (eyes / 2))
    {
        comp += (eyes / 2);
        bodies -= eyes /2;
        eyes = 0;
    }

    /* 1 eye 1 body */
    if (eyes >= bodies)
    {
        comp += bodies;
        eyes -= bodies;
        bodies = 0;
    }

    /* 2 eye, 1 body, 1 mouth*/
    if (bodies >= (eyes / 2) && mouths >= (eyes / 2))
    {
        comp += (eyes / 2);
        bodies -= comp;
        mouths -= comp;
        eyes = 0;
    }

    cout << comp << endl;
    return (0);
}

File: Contest #1/python/G-Katryoshka.py
#!/usr/bin/python3

# Taking inputs from user
user_input = input("")
user_input = user_input.split(" ")
user_input = [int(x) for x in user_input]

# Unpacking inputs
comp = 0
eyes = user_input[0]
mouths = user_input[1]
bodies = user_input[2]

# Wild cases
if eyes == 0 or bodies == 0:
    print(0)

elif eyes == 1:
    print(1)

elif bodies == 1:
    print(1)

elif bodies >= eyes and mouths >= eyes:
    print(eyes)

elif eyes >= bodies and mouths >= bodies:
    print(bodies)

# Normal cases
else:
    # 1 eye, 1 body, 1 mouth
    if bodies >= mouths and eyes >= mouths:
        comp += mouths
        eyes -= mouths
        bodies -= mouths
        mouths = 0

    if mouths >= bodies and eyes >= bodies:
        temp = bodies
        comp += int(bodies)
        eyes -= bodies
        bodies -= temp
        mouths -= temp
        temp = 0

    if mouths >= eyes and bodies >= eyes:
        comp += int(eyes)
        eyes -= comp
        bodies -= comp
        mouths -= comp

    # 2 eyes, 1 body
    if eyes >= bodies * 2:
        comp += int(bodies)
        eyes -= bodies * 2
        bodies = 0

    # 2 eyes, 1 body
    if bodies >= eyes // 2:
        comp += int(eyes // 2)
        bodies -= eyes // 2
        eyes -= eyes // 2

    print(comp)
